### PR TITLE
Password rotate action implementation.

### PR DESCRIPTION
**Pull request by averma-canonical**
# Issue
Enhancement to add action to rotate admin passwords.

# Solution
Implementation of actions get_password, set_password

# Context
Need an action to manage credentials for various admin passwords.

# Testing
Integration test added for explicit and silent password rotation scenarios.

### CODE REVIEW

**Comment by paulomach** [comment]
@averma-canonical as we talk, we have a few points to tackle:

* split configure backup into another PR. This should be about password rotation only
* remove deprecated actions for credentials retrieval
  * make sure if these are used in integrations tests, to refactor accordingly
* add system users list to `constants.py`
* if possible refactor get/set secrets without `scope`, since most probably `unit` is not used - double check legacy relations
* extend `mysql.py` lib with a update_user_password function
* replicate PR's to mysql-k8s operator

**Comment by delgod** [comment]
> * if possible refactor get/set secrets without `scope`, since most probably `unit` is not used - double check legacy relations

I would keep `scope`. Reason: we are going to implement TLS integration soon. Each unit needs its own TLS certificate but only leader can write into `scope="app"`

File: src/charm.py [hunk 1]
```diff
@@ -224,37 +218,64 @@ def _on_database_storage_detaching(self, _) -> None:
...
-        )
+    def _on_get_password(self, event: ActionEvent) -> None:
+        """Action used to retrieve the system user's password."""
+        if "username" not in event.params:
+            raise RuntimeError("Undefined parameter username.")
+
+        if event.params["username"] not in REQUIRED_USERNAMES:
+            raise RuntimeError("Invalid username.")
+
+        username = event.params["username"]
```
Comment (delgod) [inline src/charm.py:233]: I would set `username=root` as the default username to make UX simpler. and the same for setting password
```suggestion
        username = ROOT_USERNAME
        if "username" in event.params:
            username = event.params["username"]

        if username not in REQUIRED_USERNAMES:
            raise RuntimeError("Invalid username.")

```

File: src/charm.py [hunk 2]
```diff
@@ -224,37 +218,64 @@ def _on_database_storage_detaching(self, _) -> None:
...
+            raise RuntimeError("Invalid username.")
 
-    def _on_get_server_config_credentials(self, event: ActionEvent) -> None:
-        """Action used to retrieve the server config credentials."""
         event.set_results(
-            {
-                "server-config-username": SERVER_CONFIG_USERNAME,
-                "server-config-password": self._get_secret("app", SERVER_CONFIG_PASSWORD_KEY),
-            }
+            {"username": event.params["username"], "password": self._get_secret(secret_key)}
```
Comment (delgod) [inline src/charm.py:244]:
```suggestion
            {"username": username, "password": self._get_secret(secret_key)}
```

**Review by delgod — APPROVED** [review]

File: src/charm.py [hunk 3]
```diff
@@ -224,37 +220,63 @@ def _on_database_storage_detaching(self, _) -> None:
...
+            raise RuntimeError("Invalid username.")
+
+        event.set_results({"username": username, "password": self._get_secret("app", secret_key)})
+
+    def _on_set_password(self, event: ActionEvent) -> None:
+        """Action used to update/rotate the system user's password."""
+        if not self.unit.is_leader():
+            raise RuntimeError("set-password action can only be run on the leader unit.")
+
+        username = ROOT_USERNAME
```
Comment (paulomach) [inline src/charm.py:253]: same as L232

File: src/charm.py [hunk 4]
```diff
@@ -224,37 +220,63 @@ def _on_database_storage_detaching(self, _) -> None:
...
+        if username == ROOT_USERNAME:
+            secret_key = ROOT_PASSWORD_KEY
+        elif username == SERVER_CONFIG_USERNAME:
+            secret_key = SERVER_CONFIG_PASSWORD_KEY
+        elif username == CLUSTER_ADMIN_USERNAME:
+            secret_key = CLUSTER_ADMIN_PASSWORD_KEY
+        else:
+            raise RuntimeError("Invalid username.")
+
+        new_password = None
```
Comment (paulomach) [inline src/charm.py:270]: you can replace these lines with this:

```suggestion
        new_password = event.params.get("password") or generate_random_password(PASSWORD_LENGTH)
```

File: src/charm.py [hunk 5]
```diff
@@ -224,37 +220,63 @@ def _on_database_storage_detaching(self, _) -> None:
...
+        if "password" not in event.params or event.params["password"] == "":
+            new_password = generate_random_password(PASSWORD_LENGTH)
+        else:
+            new_password = event.params["password"]
+
+        current_server_config_password = self._get_secret("app", SERVER_CONFIG_PASSWORD_KEY)
```
Comment (paulomach) [inline src/charm.py:276]: you should not need to store (or pass to the function) the current password, since you are setting on the databag after changing in mysql.

File: lib/charms/mysql/v0/mysql.py [hunk 6]
```diff
@@ -979,6 +979,30 @@ def grant_privileges_to_user(
             logger.warning(f"Failed to grant privileges to user {username}@{hostname}", exc_info=e)
             raise MySQLGrantPrivilegesToUserError(e.message)
 
+    def update_user_password(
+        self, username: str, new_password: str, current_server_config_password: str
```
Comment (paulomach) [inline lib/charms/mysql/v0/mysql.py:983]: you should remove the current password setting here (see comment on `charm.py`)

File: lib/charms/mysql/v0/mysql.py [hunk 7]
```diff
@@ -979,6 +979,30 @@ def grant_privileges_to_user(
...
+    ) -> None:
+        """Updates user password in MySQL database."""
+        logger.debug(f"Updating password for {username}.")
+
+        connect_instance_address = self.instance_address
+        server_config_username = self.server_config_user
```
Comment (paulomach) [inline lib/charms/mysql/v0/mysql.py:989]: these are redundant, just use `self.*` directly on the f-strings below

**Review by paulomach — CHANGES_REQUESTED** [review]
Abi, some changes, tackle them and we are good to go

File: lib/charms/mysql/v0/mysql.py (hunk 6 above)
Comment (paulomach) [reply lib/charms/mysql/v0/mysql.py:983]: But I don't see why you would need to refresh it, i.e.: 

`actions is called -> new pass to mysql using old pass -> new pass to databag -> hook finish`

next runs of anything will run with new password already

Comment (averma-canonical) [reply lib/charms/mysql/v0/mysql.py:983]: I am talking about the scenario where 
- we rotate the serverconfig password and then after that 
- try to rotate any other password. 
 
In this scenario, the update_user_password() function which relies on the serverconfig password will fail since it's copy of the serverconfig password loaded at init would be out of sync.

File: actions.yaml [hunk 8]
```diff
@@ -1,14 +1,26 @@
...
   description: Get cluster status information
+
+get-password:
+  description: Fetch the system user's password, which is used by charm.
+    It is for internal charm users and SHOULD NOT be used by applications.
+  params:
+    username:
+      type: string
+      description: The username, the default value 'operator'.
+        Possible values - operator, backup, replication.
```
[the remark this answers is not in the document]
Comment (averma-canonical) [reply actions.yaml:14]: ack

File: actions.yaml [hunk 9]
```diff
@@ -1,14 +1,26 @@
...
+
+set-password:
+  description: Change the system user's password, which is used by charm.
+    It is for internal charm users and SHOULD NOT be used by applications.
+  params:
+    username:
+      type: string
+      description: The username, the default value 'operator'.
+        Possible values - operator, backup, replication.
```
[the remark this answers is not in the document]
Comment (averma-canonical) [reply actions.yaml:23]: ack

File: lib/charms/mysql/v0/mysql.py (hunk 7 above)
Comment (averma-canonical) [reply lib/charms/mysql/v0/mysql.py:989]: ack

File: src/charm.py [hunk 10]
```diff
@@ -224,37 +220,63 @@ def _on_database_storage_detaching(self, _) -> None:
...
-    def _on_get_server_config_credentials(self, event: ActionEvent) -> None:
-        """Action used to retrieve the server config credentials."""
-        event.set_results(
-            {
-                "server-config-username": SERVER_CONFIG_USERNAME,
-                "server-config-password": self._get_secret("app", SERVER_CONFIG_PASSWORD_KEY),
-            }
-        )
+        if "username" in event.params:
+            username = event.params["username"]
```
[the remark this answers is not in the document]
Comment (averma-canonical) [reply src/charm.py:232]: I think it will throw KeyError: 'username'

File: src/charm.py (hunk 3 above)
Comment (averma-canonical) [reply src/charm.py:253]: I think it will throw KeyError: 'username'

File: src/charm.py (hunk 4 above)
Comment (averma-canonical) [reply src/charm.py:270]: I think it will throw KeyError: 'password'

File: lib/charms/mysql/v0/mysql.py (hunk 6 above)
Comment (averma-canonical) [reply lib/charms/mysql/v0/mysql.py:983]: ack, will update the lib's copy of serverconfig password as well to keep it in sync

File: lib/charms/mysql/v0/mysql.py [hunk 11]
```diff
@@ -979,6 +979,36 @@ def grant_privileges_to_user(
...
+        try:
+            self._run_mysqlsh_script("\n".join(update_user_password_commands))
+        except MySQLClientError as e:
+            logger.exception(
+                f"Failed to update user password for user {username}",
+                exc_info=e,
+            )
+            raise MySQLCheckUserExistenceError(e.message)
+
+        if username == self.server_config_user:
```
Comment (shayancanonical) [inline lib/charms/mysql/v0/mysql.py:1009]: why is this needed? this MySQL class is instantiated anew every time a handler is invoked (since the charm class is also instantiated anew every time a handler is invoked)

File: tests/integration/test_database.py [hunk 12]
```diff
@@ -82,6 +90,131 @@ async def test_build_and_deploy(ops_test: OpsTest):
 @pytest.mark.order(2)
 @pytest.mark.abort_on_fail
 @pytest.mark.database_tests
+async def test_password_rotation(ops_test: OpsTest):
+    """Rotate password and confirm changes."""
+    num_units = len(ops_test.model.applications[DATABASE_APP_NAME].units)
+    random_unit = ops_test.model.applications[DATABASE_APP_NAME].units[num_units - 1]
```
Comment (shayancanonical) [inline tests/integration/test_database.py:96]: you could potentially index the units array by `-1` and achieve the same - but is more readable in my opinion

File: tests/integration/test_database.py [hunk 13]
```diff
@@ -82,6 +90,131 @@ async def test_build_and_deploy(ops_test: OpsTest):
...
+
+    await rotate_credentials(
+        unit=primary_unit, username=SERVER_CONFIG_USERNAME, password=new_password
+    )
+
+    updated_credentials = await fetch_credentials(random_unit, SERVER_CONFIG_USERNAME)
+    assert updated_credentials["password"] != old_credentials["password"]
+    assert updated_credentials["password"] == new_password
+
+    primary_unit = await get_primary_unit(
```
Comment (shayancanonical) [inline tests/integration/test_database.py:123]: why do we need to log what node is the primary member on line 133? could we remove lines 123-133?

File: tests/integration/test_database.py [hunk 14]
```diff
@@ -82,6 +90,131 @@ async def test_build_and_deploy(ops_test: OpsTest):
...
+    logger.debug(
+        "Test succeeded Primary unit detected before password rotation is %s", primary_unit_address
+    )
+
+    await rotate_credentials(unit=primary_unit, username=SERVER_CONFIG_USERNAME)
+
+    updated_credentials = await fetch_credentials(random_unit, SERVER_CONFIG_USERNAME)
+    assert updated_credentials["password"] != old_credentials["password"]
+
+    primary_unit = await get_primary_unit(
```
Comment (shayancanonical) [inline tests/integration/test_database.py:165]: same thing as above, do we need lines 165-175 in this test?

File: lib/charms/mysql/v0/mysql.py (hunk 11 above)
Comment (averma-canonical) [reply lib/charms/mysql/v0/mysql.py]: ack, I was not aware. was being defensive.

File: tests/integration/test_database.py (hunk 13 above)
Comment (averma-canonical) [reply tests/integration/test_database.py:122]: - primary unit is needed to issue set password action.
- it is also an actual test for the sanity of rotated password, of course post rotation

File: src/charm.py (hunk 5 above)
Comment (averma-canonical) [reply src/charm.py]: ack

File: tests/integration/test_database.py (hunk 13 above)
Comment (averma-canonical) [reply tests/integration/test_database.py]: Added comments in tests to explain what I am doing, plus changed the post rotation password check query

File: tests/integration/test_database.py (hunk 14 above)
Comment (averma-canonical) [reply tests/integration/test_database.py]: ack

**Review by shayancanonical — APPROVED** [review]

**Review by paulomach — APPROVED** [review]
